### ISSUE TITLE
systemd cannot deal with multiple mounts sharing a mount point

### ISSUE THREAD

**Issue by bl33pbl0p** [systemd/systemd#11301]
**systemd version the issue has been seen with**
> 239

**Used distribution**
 > Debian

**Expected behaviour you didn't see**
 > Stopping the mount unit of the whole stacked mount setup at a mount point.

**Unexpected behaviour you saw**
 > All mounts underneath being recursively unmounted.

**Steps to reproduce the problem**
Mount on top of some mount point where other stuff is already mounted (systemd synthesizes the unit for it since it did not go through the manager, from /proc/self/mountinfo). `systemctl stop mntpoint.mount` causes the manager to stop the mount unit but that results in everything underneath being unmounted recursively.

This is a consequence of the path -> mount unit name mapping, as in only one path can be associated with a unit at once, so it unmounts everything under that assumption.

**RFC**
My idea to overcome this limitation would be to reference count mounts on a mount point, and make the escaped path name of the mount an alias to the top level instance.

For example, if I have /bar
`mount --rbind /etc/systemd /bar`
should produce bar@1.mount, which bar.mount is aliased to.
`mount --rbind /foo /bar`
should produce bar@2.mount, which bar.mount now gets aliased to.

Perhaps beaviour on a stop job being trigged for the top level instance should be to unmount that, and nothing else underneath (unless ofcourse they need to be stopped). Comments?

**Comment by bl33pbl0p** [comment]
There is also interesting interaction between manager created mounts disappearing when stacked underneath a manual bind mount.

Consider creating a loop device /dev/loop0, attach a filesystem to it, and use systemd-mount to mount it.

This creates a mnt.mount unit bound to dev-loop0.device unit. Then `mount --rbind /etc/systemd /mnt` over it. First issue is that there is now already a discrepancy between the mount unit and the mount.

An `rmmod -f loop` operation causes the backing device to disappear, which is stacked under the bind mount, systemd rushes to stop the mount unit, and in that operation unmounts everything recursively. This is different from doing mounts manually where the top level mount remains intact (and this is noticed when one unmount the top level mount, i.e. files disappearing).

**Comment by bl33pbl0p** [comment]
```bash
% systemd-mount /dev/loop0 /mnt; mount --rbind /etc/systemd /mnt; systemd-analyze log-level debug
Dec 31 20:37:03 jupiter systemd[1]: Mounting /mnt...
Dec 31 20:37:03 jupiter systemd[1]: Mounted /mnt.
Dec 31 20:38:01 jupiter systemd[1]: Setting log level to debug.

% rmmod -f loop
Dec 31 20:38:01 jupiter systemd[1]: dev-disk-by\x2duuid-1b8aba02\x2daa1f\x2d44f2\x2d92c8\x2df6d13b44ebbc.device: Changed plugged -> dead
Dec 31 20:38:01 jupiter systemd[1]: dev-loop0.device: Changed plugged -> dead
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Trying to enqueue job mnt.mount/stop/replace
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Installed new job mnt.mount/stop as 2433
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Enqueued job mnt.mount/stop as 2433
Dec 31 20:38:01 jupiter systemd[1]: systemd-fsck@dev-loop0.service: Trying to enqueue job systemd-fsck@dev-loop0.service/stop/replace
Dec 31 20:38:01 jupiter systemd[1]: systemd-fsck@dev-loop0.service: Installed new job systemd-fsck@dev-loop0.service/stop as 2434
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Merged into installed job mnt.mount/stop as 2433
Dec 31 20:38:01 jupiter systemd[1]: systemd-fsck@dev-loop0.service: Enqueued job systemd-fsck@dev-loop0.service/stop as 2434
Dec 31 20:38:01 jupiter systemd[1]: sys-devices-virtual-block-loop0.device: Changed plugged -> dead
Dec 31 20:38:01 jupiter systemd[1]: sys-devices-virtual-block-loop0.device: Collecting.
Dec 31 20:38:01 jupiter systemd[1]: dev-disk-by\x2duuid-1b8aba02\x2daa1f\x2d44f2\x2d92c8\x2df6d13b44ebbc.device: Collecting.
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2duuid_2d1b8aba02_5cx2daa1f_5cx2d44f2_5cx2d92c8_5cx2df6d13b44ebbc_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7880 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2duuid_2d1b8aba02_5cx2daa1f_5cx2d44f2_5cx2d92c8_5cx2df6d13b44ebbc_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7881 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=7882 reply_cookie=0 signature=so error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7883 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7884 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=7885 reply_cookie=0 signature=so error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7886 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7887 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobNew cookie=7888 reply_cookie=0 signature=uos error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobNew cookie=7889 reply_cookie=0 signature=uos error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: systemd-logind.service: Got notification message from PID 1019 (WATCHDOG=1)
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: About to execute: /bin/umount /mnt -c
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Forked /bin/umount as 14010
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Changed mounted -> unmounting
Dec 31 20:38:01 jupiter systemd[1]: Unmounting /mnt...
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mnt_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7890 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mnt_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7891 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/job/2433 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7892 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: systemd-journald.service: Got notification message from PID 855 (FDSTORE=1)
Dec 31 20:38:01 jupiter systemd[1]: systemd-journald.service: Added fd 79 (n/a) to fd store.
Dec 31 20:38:01 jupiter systemd[1]: libmount event [rescan: yes]
Dec 31 20:38:01 jupiter systemd[1]: dev-loop0.device: Changed dead -> tentative
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Trying to enqueue job mnt.mount/start/fail
Dec 31 20:38:01 jupiter systemd[1]: Requested transaction contradicts existing jobs: Resource deadlock avoided
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mnt_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7893 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mnt_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7894 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7895 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7896 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dmapper_2dvg_5cx2droot_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7897 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dmapper_2dvg_5cx2droot_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7898 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 79 (stored), closing.
Dec 31 20:38:01 jupiter systemd[1]: Received SIGCHLD from PID 14010 (umount).
Dec 31 20:38:01 jupiter systemd[1]: Child 14010 (umount) died (code=exited, status=0/SUCCESS)
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Child 14010 belongs to mnt.mount.
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Mount process exited, code=exited status=0
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Mount still present, trying again.
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: About to execute: /bin/umount /mnt -c
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Forked /bin/umount as 14013
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mnt_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7899 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mnt_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7900 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: systemd-journald.service: Got notification message from PID 855 (FDSTORE=1)
Dec 31 20:38:01 jupiter systemd[1]: systemd-journald.service: Added fd 79 (n/a) to fd store.
Dec 31 20:38:01 jupiter systemd[1]: libmount event [rescan: yes]
Dec 31 20:38:01 jupiter systemd[1]: dev-loop0.device: Changed tentative -> dead
Dec 31 20:38:01 jupiter systemd[1]: systemd-fsck@dev-loop0.service: Trying to enqueue job systemd-fsck@dev-loop0.service/stop/replace
Dec 31 20:38:01 jupiter systemd[1]: systemd-fsck@dev-loop0.service: Merged into installed job systemd-fsck@dev-loop0.service/stop as 2434
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Merged into installed job mnt.mount/stop as 2433
Dec 31 20:38:01 jupiter systemd[1]: systemd-fsck@dev-loop0.service: Enqueued job systemd-fsck@dev-loop0.service/stop as 2434
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7901 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7902 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dmapper_2dvg_5cx2droot_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7903 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dmapper_2dvg_5cx2droot_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7904 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/job/2433 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7905 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/job/2434 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7906 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 79 (stored), closing.
Dec 31 20:38:01 jupiter systemd[1]: Received SIGCHLD from PID 14013 (umount).
Dec 31 20:38:01 jupiter systemd[1]: Child 14013 (umount) died (code=exited, status=0/SUCCESS)
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Child 14013 belongs to mnt.mount.
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Mount process exited, code=exited status=0
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Changed unmounting -> dead
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Job mnt.mount/stop finished, result=done
Dec 31 20:38:01 jupiter systemd[1]: Unmounted /mnt.
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=7907 reply_cookie=0 signature=uoss error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: mnt.mount: Collecting.
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mnt_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7908 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mnt_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7909 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=7910 reply_cookie=0 signature=so error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: systemd-fsck@dev-loop0.service: Changed exited -> dead
Dec 31 20:38:01 jupiter systemd[1]: systemd-fsck@dev-loop0.service: Job systemd-fsck@dev-loop0.service/stop finished, result=done
Dec 31 20:38:01 jupiter systemd[1]: Stopped File System Check on /dev/loop0.
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=7911 reply_cookie=0 signature=uoss error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: systemd-fsck@dev-loop0.service: Collecting.
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/systemd_2dfsck_40dev_2dloop0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7912 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/systemd_2dfsck_40dev_2dloop0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7913 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=7914 reply_cookie=0 signature=so error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: systemd-quotacheck.service: Collecting.
Dec 31 20:38:01 jupiter systemd[1]: dev-loop0.device: Collecting.
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=7915 reply_cookie=0 signature=so error-name=n/a error-message=n/a
Dec 31 20:38:01 jupiter systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=7916 reply_cookie=0 signature=so error-name=n/a error-message=n/a
```
https://pastebin.com/raw/Bsyqjx2C

**Comment by poettering** [comment]
So, while the kernel allows mounting multiple file systems onto the same mount point, what's the usecase for that? systemd has a simpler model there, and cannot be used to establish multiple mounts onto the same mount point, and will ensure that if something was a mount point before and you stop it that after it isn't a mount point anymore.

I mean, this logic actually matches the whole recursive mount logic of systemd as a whole: if you have a a mount point /a and below /a/b and you tell systemd to umount /a then it will implicitly umount /a/b first too. Hence, it's only natural if systemd also unmounts all instances of /a on top of each other too if you ask it to unmount /a...

Hence, I think the current behaviour makes a ton of sense, and I fail to see the benefit of baking a concept of multiple mounts on the same mount point... 

Hence, enlighten me, why should systemd care?

**Comment by bl33pbl0p** [comment]
It appears to me as more of a relic of mapping paths directly to unit names than anything else, which is certainly not true in case of Linux. It cannot keep track of which one of those instances was its own, so it is under the assumption the unit owns the mountpoint and is a singleton, loops when it sees mount points underneath still exist (treating it as some error) until everything goes away, cleaning it up.

The case you talk about has been explicitly asked for by the user, and I agree that it is correct. I still think after it traverses down to /a after unmounting everything on /b, it should perhaps unmount everything. With the refcounting logic however, I can see what unit a.mount is aliased to and then ask it explicitly to only stop a@3.mount, so that the other two stacked underneath remain (and I expect it to enqueue a stop job only for a@2.mount losing its backing device only after a@3.mount goes away, because that mount is perfectly usable until then).

There is breakage when you manually mount something over a path, but don't create a mount unit for it (or generate one), PID1 internally synthesizes one for the path mapping it to a unit name, and one cannot use mount units/systemd-mount over it.

I already gave a example of where it breaks user mounts, if I bind mount something on top of a mountpoint systemd started a mount unit on (and say have it backed by a USB drive), removing the USB will cause systemd remove my bind mount before it gets to stopping the mount unit (and recursively everything underneath it) because the device unit it was bound to is gone. This is certainly not desirable, and not how things worked before (and I don't think they were "broken" before, the top level mount is still perfectly usable as long as it is there).

The refcounting logic might allow it to track which instance was its own, and when doing a tranversal down the depth to see if the next unit still has a backing device and is active, and then do nothing to it. I think it would fit very neatly into the whole logic, as now manual mounts can be controlled through systemd and vice versa. Currently this is not the case.

Maybe I should probably use a script if this is nothing to fix in systemd then, perhaps, and never ask it to act on those mounts.

**Comment by tomty89** [comment]
I am not sure if I missed something, but I don't think a refcount is not enough for what you proposed, but something much more complexed.

Say you have /foo mounted twice and /foo/bar mounted once, if you want to know when /foo/bar should be unmounted, you will need to know when it was mounted (before or after the second /foo), that means the the dependencies of the mount units will need to be at instance-level. Similar goes for the dependencies between device units and mount unit instances.

Only with such instance-level dependencies you can then start to care about refcount and determine whether the stopping of an instance should only be "queued" or "done". It would be ridiculous if systemd only start doing the jobs until the refcount(s) turns zero, wouldn't it? Coz with the example above, it makes no sense to "hold" foo@2.mount and bar@1.mount if /foo/bar is mounted after /foo is mounted twice, does it?

**Comment by bl33pbl0p** [comment]
Eh, the refcount thing is just an example. I am not proposing something, it was like what struck me first.

Anyway, what you talk about isn't that hard. Since the top level mount gets aliased to the escaped path name mount unit, it would work just fine. How systemd should deal with it is totally orthogonal to that.

If I mount something at /foo first, it becomes foo@1.mount aliased to foo.mount, now if I mount something at /foo/bar, it becomes foo-bar@1.mount aliased to foo-bar.mount, if you now mount something else on top of /foo, it will mask the /foo/bar mount, so it is obvious foo@2.mount now points to foo.mount, with nothing on top. Ofcourse, since foo-bar@1.mount was on top of foo@1.mount, it gains a Requires= and After= on it, so that stopping foo@1.mount stops it before.

However, when I ask systemd to umount /foo recursively by asking it to stop foo.mount, it will then walk down recursively, and the dependencies will naturally cause foo-bar@1.mount to have a stop job in queue before the one for foo@1.mount get processed, by walking down RequiredBy= in it and waiting on it.

The second case is about not acting on the BindsTo= device unit disappearing for a mount unit stacked underneath. That is a contradictary case indeed, in that there is no way to wait on some other unit until it has a job, and since we don't want the top level mount to go away, waiting on it to have a stop job becomes impossible. All of this makes the mount state machine pretty complex. However, this is why I have the InterruptedBy= patch. 

I envision that systemd actually removes the BindsTo= dep on the lower stacked mount as soon as it sees the device disappearing, and instead adds an Interrupts= dep to the mount above it! That causes that mount to trigger a stop job for the lower stacked mount unit as soon as it goes to inactive, cleaning stuff up nicely. It could probably do it when a stop job is triggered too, like splitting PartOf= into two lower level directives, one enabling stop job propagation, and one restart job propagation (because restart job propagation doesn't make sense here).

Anyway, this probably has some other corner cases, but it's something that came to my mind first, fits in well with the whole dependency concept. Ofcourse I'd not oppose if this is solved in some other manner that makes things even simpler.

**Comment by tomty89** [comment]
I think you have missed my point. While you're focusing on what should happen when foo@1.mount or foo@2.mount is being stopped deliberately (by you or the removal of their respective device or whatever), you forget foo-bar@N.mount can be deliberately stopped and you need to care about that, and that is where the story gets complex (I'm not saying that it's infeasible though).

As I said, you need to check whether foo-bar@N.mount depends on the latest /foo instance. It then also means that, whenever there is a refcount decrement, you need to recursively check whether it and anything depends on it have a queued stop job, as foo-bar@N might depends on foo@1.mount out of two /foo (or even worse).

I am not trying to talk you off this idea (well at least not entirely), but since you suggest refcount, I tell the bigger picture I see here (and it may not even be the whole picture).

But yeah personally I don't like it a lot either, coz we are talking about putting something very essential (the handling of mounts, that is) at risk of being broken for some corner case (which I don't really see a concrete/sensible use-case here). And things like putting a stop job on hold "forever" makes me shiver.

**Comment by bl33pbl0p** [comment]
That's easy, just patch in a RefuseManualStop= or so, so that only jobs triggered by dependency propagation have any effect (and remove it again when it becomes the top level mount).

Also, nobody is waiting on the stop job forever! We would just be changing the direction it propagates from, instead of the device, it becomes the mount unit on top of it that causes it to be cleaned up (as that is only the sensible way to do so). Anyway, feel free to discuss this on IRC, you know where to find me.

Also, I think things are already broken badly with the current status quo, so \*something\* needs to be done, anyway.

**Comment by tomty89** [comment]
> That's easy, just patch in a RefuseManualStop= or so, so that only jobs triggered by dependency propagation have any effect

It wouldn't work. I mean, it's same difference (or perhaps inferior) of what I said.

> (and remove it again when it becomes the top level mount).

As you can only tell it (e.g. foo-bar@1.mount) is a top-level mount if all its "parents" (e.g. foo@1.mount) are top-level mounts. So it would be also a recursive (only backward) check on refcounts as well.

The worse part of this approach is, it ignores the fact that the deliberate stop can be caused by removal of device.

You see, the problem isn't that it's difficult to implement, but rather, such multiple mounts cases themselves are complex. What you suggested, refcount that is, is probably not just an approach of how we can make systemd do it, but simply part of (but only part of) the cases itself, even when systemd is not involved.

Say /foo/bar is mounted under/before the second /foo, then if you *need* /foo/bar be unmounted, you will either have to `umount /foo` (the second /foo) then `umount /foo/bar`, or if you for some reason don't want to `umount /foo` yet, you'll have to *remember* to `umount /foo/bar` once you can and have done `umount /foo`.

When systemd becomes *you*, isn't *remember* doomed to become the "forever" wait for (or pending of) the stop job? (When the cause is removal of device, wouldn't *need* become the case?) And the fact that *you* know when (as in, the order) to `umount /foo/bar` becomes refcounts (and the checking/tracking of them).

P.S. I would rather keep the discussion here, as the reason I am making comments for this is, in case anyone is really gonna do it, at least he will have a chance of seeing the traps I am already seeing. It will be easier to keep track of whether I have missed anything or made any mistakes as well. (So only if you are starting to feel like, you just want someone to persuade you to close this, then we should keep it between us via IRC :P)

**Comment by bl33pbl0p** [comment]
Well, top level means the instance that should be considered the mount unit mapping to the escaped path name. There are no levels between foo@1 and foo-bar@1 unless you mount something *over* foo@1, but that still keeps foo-bar@1 dependent on foo@1 to be around (and its device).

Yes, they can be stopped when the bound device goes away, which is why for anything that is bound by a device but not aliased to a mount unit escaped by the path name, when that happens, systemd removes the dep, adds an equivalent InterruptedBy= and stop equivalent of PartOf= too foo@n+1 if it is foo-bar@n, and just stays there, and adds a RefuseManualStop, so you cannot screw it. Then, when you trigger a stop job for foo@n+1, it propagates that too foo-bar@n, so that it gets cleaned up as its device is gone (but only when the upper one is going away). foo@n is still there as it is still valid. If you request foo.mount to stop, then it recursively goes to n=1 and unmounts everything, and doing that should walk RequiredBy= for any paths that are nested inside them (so reachinf foo@1 and stopping would walk RequiredBy=foo-bar@1 and stop it).

So, to summarise, it works like this:

* You mount something on /foo, a foo@1.mount is created, and foo.mount is aliased to this unit.
* You mount something on /foo/bar, a foo-bar@1.mount is created, and foo-bar.mount is aliased to this unit. systemd will set Requires= + After= for foo@1.mount (not foo.mount) in this unit, causing that unit to gain a symmetric Before= and RequiredBy=.
* You mount another thing on /foo, now, foo@2.mount is created, and foo.mount is aliased to this unit. The alias for foo-bar.mount is destroyed (and will be recreated once we're back down there).
* You remove the backing device for foo-bar@1.mount. systemd will instead, noticing that the escaped path mount unit is not pointing to this unit, indicating is stacked underneath something, will drop the BindsTo= dependency immediately, and add a hypothetical InterruptedBy= + PartOf= equivalent for just stop to whatever foo@n.mount unit it needs, + 1, so foo@2.mount in this case (as it can track there is a mount on top internally, already, since the foo.mount alias is active). This ensured our nested mount unit remains active only until the mount masking it is up.
It also adds a RefuseManualStop= for this foo-bar@1.mount so you cannot trigger a stop accidently.

Now, you can do two things.

* You stop foo@2.mount explicitly by reading the Names= property of foo.mount, which will cause foo-bar@1.mount to get cleaned up, but foo@1.mount to stay, as it is still active.
* You stop foo.mount, which causes a recursive unmount of that path, first stopping foo@2.mount, which causes foo-bar@1.mount to be cleaned up, then stopping foo@1.mount, which now leads to the path fully getting cleaned up.

What am I missing?

**Comment by bl33pbl0p** [comment]
Also, I really hope that something like InterruptedBy= goes in, as it would be needed for something like this, and PartOf= is split into a stop job propagator and a restart job propagator (for things like the above, restart propagation makes no sense at all). This goes one step closer to making the job engine fully generic when it comes to propagation.

For job engine:
Wants= : start job propagation
Wants= + Requisite= : start job propagation + invalidation like Requires=, to be used with After=
PartOf= : stop+restart job propagation

For unit transition without involvement of manager:
InterruptedBy= : stop job propagation on reaching inactive/failed

Also,
Wants= + Requisite= + PartOf= + InterruptedBy= becomes equivalent to BindsTo=, except the strong unit_check_bindsto and unit_verify_deps checks (which actually warrants a stronger version of Wants= like Pulls= or so but without stop/restart job propagation, as Requires= is already done with Wants= + Requisite= (+ After=)).

**Comment by poettering** [comment]
@AdrianVovk not sure I follow, but your /usr is already a mount point? from where? why do you not simply replace that mount?

**Comment by AdrianVovk** [comment]
`/usr` comes from the system's root (from OSTree, which does some black magic that I don't understand but I'm sure @cgwalters can elaborate if necessary). No `usr.mount` unit exists on the system. Here's some relevant outputs:
```
[~]$ cat /proc/self/mountinfo | grep /usr
65 63 8:17 /ostree/deploy/carbonOS/deploy/ef7cb496debca38edd82c7bd15990837f188212c8128502d107b20035270ed08.0/usr /usr ro,relatime shared:3 - ext4 /dev/sdb1 rw
144 65 0:37 / /usr rw,relatime shared:74 - overlay overlay ro,lowerdir=/usr:/var/apps,index=off,xino=off
[~]$ systemctl cat usr.mount
No files found for usr.mount.
[~]$ systemctl status usr.mount
● usr.mount - /usr
   Loaded: loaded (/proc/self/mountinfo)
   Active: active (mounted)
    Where: /usr
     What: overlay
```

If I remember correctly, if my overlay isn't mounted the only difference is `What: /dev/sdb1`. I can't give you a log because my GUI depends on the overlay being active. 

If I try to create a `usr.mount` unit that does the overlay, starting it will succeed but running `systemctl status usr.mount` still prints  `What: /dev/sdb1`.

**Comment by cgwalters** [comment]
@AdrianVovk You may be interested in https://github.com/projectatomic/rpm-ostree/issues/639
At some point I'd like to land some of the "live changes" support in libostree.

(TL;DR rpm-ostree's support for this today just writes "underneath" the read-only `/usr` mount; in the end ostree is just plain files, don't strictly need overlayfs although it can help make some parts simpler)

**Comment by AdrianVovk** [comment]
@cgwalters What are the downfalls of overmounting an overlayfs? It seems quite easy to just install things into `/var/apps` and then overlay it on top of `/usr`. Why should anything manipulate "underneath" the read-only mount?

**Comment by bl33pbl0p** [comment]
BTW, finding a solution for https://github.com/systemd/systemd/issues/10133 should probably be done with this one, in that case, the count mounts over one point thing probably needs to be reworked a bit.

**Comment by apteryks** [comment]
Any workaround to this? I guess using a more general .service to do the work can be used as an inelegant workaround? My use case is installing Guix in environments where the root file system is too small; in this situation you might want to bind mount `/gnu/store` from somewhere else (from a larger storage device) say. There is a `gnu-store.mount`installed by Guix on systemd systems that remounts `/gnu/store` as read-only; the main service, `guix-daemon.service` briefly stops `gnu-store.mount` when it launches to be able to gain write access to it. But this doesn't work as stopping `gnu-store.mount` also unmounts the underlying bind mount, leaving you with an empty `/gnu/store`.
